xray/aws.go

```diff
@@ -96,8 +96,12 @@ var xRayAfterSendHandler = request.NamedHandler{
 		if curseg != nil && curseg.Name == "attempt" {
 			// An error could have prevented the connect subsegment from closing,
 			// so clean it up here.
-			for _, subsegment := range curseg.rawSubsegments {
-				if subsegment.Name == "connect" && subsegment.safeInProgress() {
+			curseg.RLock()
+			s := curseg.rawSubsegments
```

Comment:
Why we create a copy here instead of directly using it?

Reply:
With regards to optimization, in the sense, we hit this code for every AWS SDK API call being traced. Since we are iterating over `rawSegments`, its better to have copy, so another Go routine, if present, can update same `Segment` with minimum lock overhead.

Reply:
Sure,
This code is executed for AWS API call, traced by the AWS SDK handler. For a given AWS SDK call we have possibility of generating `connect` subsegment inside `attempt` subsegment. The structure looks like :
```
SegA
|-Dynamo(SubsegA)
|-|-attempt (SubSegB)
|---|----connect(SubSegC)
|---|----request(SubSegD)
|---|----response(SubSegE)
```
The above code generates a copy for all children of `attempt` subsegment and closes `connect` subsegment. Note that [rawSegments](https://github.com/aws/aws-xray-sdk-go/blob/master/xray/segment_model.go#L69) is a array of pointers to `Segment` instances. Hence even after we make copy, the locks held by this copy are on the same instances of `Subsegment` and are valid inside the loop. 

 `SubSegB.rawSegments` should be thread safe since another thread may populate children of `attempt` subsegment. 
The unit test written for aws sdk call produces a data race, where we are iterating children of `attempt` subsegment, however, `response` subsegment is concurrently populated leading to data race at `subsegment.Name` and `SubSegB.rawSegments`. Hence the access to name of the subsegment is made thread safe by calling `getName()` method and making a copy respectively. 

*Note* :`s := curseg.rawSubsegments` is not a deep copy . Hence in the updated revision, I have made a deep copy to prevent any data races w.r.t `SubSegB.rawSegments` being populated by another go routine while iterating in this code block.